erpnext/payroll/doctype/salary_slip/salary_slip.py

```diff
@@ -507,6 +525,118 @@ def add_structure_components(self, component_type):
+		processed_overtime_slips = []
+		overtime_types_details = {}
+		for slip in overtime_slips:
+			holiday_date = self.get_holidays_for_employee(slip.from_date, slip.to_date, as_dict=1)
+
+			holiday_date_map = {}
+			for date in holiday_date:
+				holiday_date_map[cstr(date.holiday_date)] = date
+
+			details = self.get_overtime_details(slip.name)
+
+			for detail in details:
+				overtime_hours = detail.overtime_duration / 3600
+
+				if detail.overtime_type not in overtime_types_details:
+					details, applicable_components = self.get_overtime_type_detail(detail.overtime_type)
+					overtime_types_details[detail.overtime_type] = details
+					if len(applicable_components):
+						overtime_types_details[detail.overtime_type]["components"] = applicable_components
+					else:
+						frappe.throw(_("Select applicable components in Overtime Type: {0}").format(
+							frappe.bold(detail.overtime_type)))
+
+					if "applicable_amount" not in overtime_types_details[detail.overtime_type].keys():
+						component_amount = sum([data.default_amount for data in self.earnings if data.salary_component in
+							overtime_types_details[detail.overtime_type]["components"] and not data.get('additional_salary', None)])
+
+						overtime_types_details[detail.overtime_type]["applicable_daily_amount"] = component_amount/self.total_working_days
+
+				standard_working_hours = detail.standard_working_time/3600
+				applicable_hourly_wages = overtime_types_details[detail.overtime_type]["applicable_daily_amount"]/standard_working_hours
+
+				weekend_multiplier = overtime_types_details[detail.overtime_type]['standard_multiplier']
+				public_holiday_multiplier = overtime_types_details[detail.overtime_type]['standard_multiplier']
+
+				if overtime_types_details[detail.overtime_type]['applicable_for_weekend']:
+					weekend_multiplier = overtime_types_details[detail.overtime_type]['weekend_multiplier']
+				if overtime_types_details[detail.overtime_type]['applicable_for_public_holiday']:
+					public_holiday_multiplier = overtime_types_details[detail.overtime_type]['public_holiday_multiplier']
+				overtime_date = cstr(detail.date)
+				if overtime_date in holiday_date_map.keys():
+					if holiday_date_map[overtime_date].weekly_off == 1:
+						calculated_amount = overtime_hours * applicable_hourly_wages * weekend_multiplier
+						weekends_duration_amount += calculated_amount
+					elif holiday_date_map[overtime_date].weekly_off == 0:
+						calculated_amount = overtime_hours * applicable_hourly_wages * public_holiday_multiplier
+						public_holidays_duration_amount += calculated_amount
+				else:
+					calculated_amount = overtime_hours * applicable_hourly_wages *\
+						overtime_types_details[detail.overtime_type]['standard_multiplier']
+					standard_duration_amount += calculated_amount
+
+			processed_overtime_slips.append(slip.name)
+
+		return [weekends_duration_amount, public_holidays_duration_amount, standard_duration_amount] , processed_overtime_slips
```

Comment:
This function is too big! Very difficult to read/debug. Break it down into smaller 7-10 line functions.

Also, this is not working. I have an Overtime Slip:

![image](https://user-images.githubusercontent.com/24353136/126977431-8ef44947-00fb-4bae-b59e-2453deae1b9e.png)

Not reflecting in Salary Slip (Overtime Allowance salary component not added automatically)

![image](https://user-images.githubusercontent.com/24353136/126977491-ad1d9163-7431-4afa-aa6e-c4014d8a725c.png)

Reply:
Done 
![image](https://user-images.githubusercontent.com/32095923/129007270-1d019452-b0cf-450b-987a-85d3566bda43.png)